File: src/desktop/modern/src/application/services/positioning/arrow_location_calculator.py
"""
Arrow Location Calculator Service - Working Implementation

This service calculates arrow locations based on motion data.
"""

from typing import Any, Optional, Protocol, runtime_checkable
from domain.models.core_models import MotionData, MotionType, Location


@runtime_checkable
class IArrowLocationCalculator(Protocol):
    """Interface for arrow location calculator."""

    def calculate_location(self, motion: MotionData) -> Location:
        """Calculate arrow location for given motion."""
        ...


class ArrowLocationCalculator:
    """Arrow location calculator service implementation."""

    def __init__(self):
        """Initialize the service."""
        pass

    def calculate_location(self, motion: MotionData) -> Location:
        """Calculate arrow location based on motion data."""
        if motion.motion_type == MotionType.STATIC:
            return motion.start_loc
        elif motion.motion_type == MotionType.DASH:
            # For now, return start location - would delegate to dash service in real implementation
            return motion.start_loc
        elif motion.motion_type in [MotionType.PRO, MotionType.ANTI]:
            # Calculate shift location (simplified logic)
            return self._calculate_shift_location(motion.start_loc, motion.end_loc)
        else:
            return motion.start_loc
    
    def _calculate_shift_location(self, start_loc: Location, end_loc: Location) -> Location:
        """Calculate shift location between two positions."""
        # Simplified shift calculation logic
        shift_map = {
            (Location.NORTH, Location.EAST): Location.NORTHEAST,
            (Location.EAST, Location.SOUTH): Location.SOUTHEAST,
            (Location.SOUTH, Location.WEST): Location.SOUTHWEST,
            (Location.WEST, Location.NORTH): Location.NORTHWEST,
            (Location.NORTHEAST, Location.NORTHWEST): Location.NORTH,
            (Location.NORTHEAST, Location.SOUTHEAST): Location.EAST,
            (Location.SOUTHWEST, Location.SOUTHEAST): Location.SOUTH,
            (Location.NORTHWEST, Location.SOUTHWEST): Location.WEST,
        }
        
        return shift_map.get((start_loc, end_loc), start_loc)


# Export the service class and interface
__all__ = ["ArrowLocationCalculator", "IArrowLocationCalculator"]

File: src/web/src/lib/components/pictograph/utils/arrowLocationCalculator.ts
/**
 * Arrow Location Calculator
 *
 * Determines arrow location based on start and end positions using the same logic
 * as the desktop app's ShiftLocationCalculator.
 */

export interface ArrowLocationInput {
  start_loc: string;
  end_loc: string;
  motion_type: string;
}

export function calculateArrowLocation(input: ArrowLocationInput): string {
  const { start_loc, end_loc, motion_type } = input;

  // Return empty string if motion state is not initialized
  if (!start_loc || !end_loc) {
    return "";
  }

  // For PRO, ANTI, FLOAT - use shift location calculation
  if (["pro", "anti", "float"].includes(motion_type)) {
    return calculateShiftLocation(start_loc, end_loc);
  }

  // For other types (DASH, STATIC), return start location
  return start_loc;
}

function calculateShiftLocation(start_loc: string, end_loc: string): string {
  const directionPairs: Record<string, string> = {
    // Diagonal combinations
    "n,e": "ne",
    "e,n": "ne",
    "e,s": "se",
    "s,e": "se",
    "s,w": "sw",
    "w,s": "sw",
    "w,n": "nw",
    "n,w": "nw",

    // Corner to cardinal combinations
    "ne,nw": "n",
    "nw,ne": "n",
    "ne,se": "e",
    "se,ne": "e",
    "sw,se": "s",
    "se,sw": "s",
    "nw,sw": "w",
    "sw,nw": "w",
  };

  const key1 = `${start_loc},${end_loc}`;
  const key2 = `${end_loc},${start_loc}`;

  return directionPairs[key1] || directionPairs[key2] || "";
}

// Constants matching the desktop app
export const LOCATIONS = {
  NORTH: "n",
  EAST: "e",
  SOUTH: "s",
  WEST: "w",
  NORTHEAST: "ne",
  SOUTHEAST: "se",
  SOUTHWEST: "sw",
  NORTHWEST: "nw",
} as const;

export const MOTION_TYPES = {
  PRO: "pro",
  ANTI: "anti",
  FLOAT: "float",
  DASH: "dash",
  STATIC: "static",
} as const;
